src/DGtal/topology/CubicalComplex.h

```diff
@@ -0,0 +1,1307 @@
+       * @param it an iterator pointing on a cell of the complex.
+       */
+      Iterator( Alias<CubicalComplex> cc, Dimension d, 
+                CellMapIterator it )
+        : myCC( &cc ), myD( d ), myIt( it )
+      {
+        ASSERT( d <= myCC->dimension );
+        myItEnd = myCC->end( d );
+        nextDimension();
+      }
+
+    private:
+      friend class boost::iterator_core_access;
+
+      void nextDimension() 
+      {
+        while ( myIt == myItEnd )
+          {
+            if ( ++myD > myCC->dimension ) break;
+            myIt    = myCC->begin( myD );
+            myItEnd = myCC->end( myD );
+          }
+      }
+
+      void increment()
+      {
+        ASSERT( myCC != 0 );
+        ++myIt;
+        nextDimension();
+      }
+
+      bool equal( const Iterator& other ) const
+      { 
+        return ( myD == other.myD ) && ( myIt == other.myIt );
+      }
+
+      Cell const& dereference() const
+      { 
+        return myIt->first;
+      }
+
+      Dimension dimension() const
+      { 
+        return myD;
+      }
+      
+    private:
+      CubicalComplex* myCC;
+      Dimension myD;
+      CellMapIterator myIt;
+      CellMapIterator myItEnd;
+    };
+
+    // ----------------------- STL inner types ------------------------------
+  public:
+
+    // Renaming for STL-type of iterator.
+    typedef ConstIterator                           const_iterator;
+    typedef Iterator                                iterator;
+    typedef Cell                                    value_type;
+    typedef Cell const&                             reference;
+    typedef Cell const&                             const_reference;
+    typedef typename CellContainer::size_type       size_type; 
+    typedef typename CellContainer::difference_type difference_type;
+    typedef Cell const*                             pointer;
+    typedef Cell const*                             const_pointer;
+
+    // ----------------------- Standard services ------------------------------
+  public:
+
+    /**
+    * Destructor.
+    */
+    ~CubicalComplex();
+
+  protected:
+    /**
+    * Constructor. The cubical complex is not valid. A user may not
+    * instantiate an empty CubicalComplex, because it needs the
+    * Khalimsky space.
+    */
+    CubicalComplex();
+  public:
+
+    /**
+    * Constructor of empty complex. Needs a space to represents
+    * cubical cells.
+    *
+    * @param aK a Khalimsky space.
+    */
+    CubicalComplex ( ConstAlias<KSpace> aK );
+
+    /**
+    * Copy constructor.
+    * @param other the object to clone.
+    * Forbidden by default.
+    */
+    CubicalComplex ( const CubicalComplex & other );
+    
+    /**
+    * Constructor a complex from a digital set.
+    * @param set - a digital set from which to create a complex. 
+    * Set has to be of the same dimension as a Khalimsky space.
+    */
+    template < typename TDigitalSet >
+    void construct ( const TDigitalSet & set );
```

Comment:
This was added by @copyme .
I propose to keep it here and to do a helper with more construct in a separate PR.